File: tests/verify/observer.py
"""Independent zor observation of a real fux pane, plus failure isolation of control clients.

Run as: observer.py FUX_BIN ZOR_BIN. Zor is started explicitly by this harness (as a user would),
consumes fux's documented control surface (`FUXCTL2` preface, `list`, `capture`) and emits OSC
7877 reports on its stdout. Fux never spawns, supervises or parses observers.
"""
import json
import os
import re
import signal
import socket
import subprocess
import sys
import tempfile
import time
from pathlib import Path

FUX = str(Path(sys.argv[1]).resolve())
ZOR = str(Path(sys.argv[2]).resolve())


def rpc(path, value, preface=b'FUXCTL2\n'):
    with socket.socket(socket.AF_UNIX) as stream:
        stream.settimeout(3)
        stream.connect(str(path))
        stream.sendall(preface)
        received = b''
        while len(received) < 8:
            chunk = stream.recv(8 - len(received))
            if not chunk:
                raise EOFError(received)
            received += chunk
        assert received == b'FUXCTL2\n', received
        stream.sendall(json.dumps(value).encode() + b'\n')
        output = b''
        while b'\n' not in output:
            chunk = stream.recv(8192)
            if not chunk:
                raise EOFError(output)
            output += chunk
            assert len(output) <= 1024 * 1024
        return json.loads(output)


def until(function, description, timeout=10):
    end = time.monotonic() + timeout
    while time.monotonic() < end:
        value = function()
        if value:
            return value
        time.sleep(.05)
    raise AssertionError(description)


with tempfile.TemporaryDirectory(prefix='fo-', dir='/tmp') as directory:
    root = Path(directory)
    config = root / 'config/fux'
    config.mkdir(parents=True)
    rules = root / 'rules'
    rules.mkdir()
    (rules / 'test.toml').write_text(
        "id='test'\nprompt_marker='>'\nblock_markers=[]\n"
        "[[rules]]\nid='working'\nstate='working'\nregion='progress'\ncontains=['1:50']\nvisible_working=true\n"
        "[[rules]]\nid='idle'\nstate='idle'\nregion='title'\ncontains=['OBS_IDLE']\nvisible_idle=true\n")
    command = ("stty raw -echo; printf 'READY\\033]9;4;1;50\\007'; dd bs=1 count=1 >/dev/null 2>&1; "
               "printf '\\033[2J\\033[HIDLE\\033]9;4;0;0\\007\\033]2;OBS_IDLE\\007'; sleep 60")
    (config / 'config.toml').write_text('default-command = { argv = ' + json.dumps(['/bin/sh', '-c', command]) + ' }\n')
    env = os.environ.copy()
    env.update(HOME=directory, XDG_RUNTIME_DIR=directory, XDG_STATE_HOME=directory + '/state',
               XDG_CONFIG_HOME=directory + '/config', SHELL='/bin/sh')
    server = subprocess.Popen([FUX, 'serve'], env=env, stdout=subprocess.DEVNULL, stderr=subprocess.PIPE)
    control = root / 'fux/default.sock'
    observer = None
    try:
        until(control.exists, 'control socket')

        def pane():
            return rpc(control, {'command': 'list', 'id': 1})['result']['value']['workspaces'][0]['tabs'][0]['panes'][0]

        def capture():
            return rpc(control, {'command': 'capture', 'id': 1, 'pane': 1, 'attrs': False, 'scrollback': 0, 'max_bytes': 4096})['result']['value']['text']

        first = pane()
        pid = first['pid']
        until(lambda: 'READY' in capture(), 'pane output visible through capture')
        until(lambda: pane().get('progress') == [1, 50], 'progress report surfaces in the listing')

        # A malformed control client is rejected without disturbing the pane.
        with socket.socket(socket.AF_UNIX) as bad:
            bad.settimeout(3)
            bad.connect(str(control))
            bad.sendall(b'FUXCTL1\n')
            try:
                reply = bad.recv(8)
            except (ConnectionResetError, BrokenPipeError):
                reply = b''
            assert reply in (b'FUXCTL2\n', b''), reply
        with socket.socket(socket.AF_UNIX) as stream:
            stream.settimeout(3)
            stream.connect(str(control))
            stream.sendall(b'FUXCTL2\n')
            stream.recv(8)
            stream.sendall(b'{"command":"popup","id":5,"argv":["x"]}\n')
            reply = json.loads(stream.recv(65536).split(b'\n')[0])
            assert reply['status'] == 'failed' and reply['error']['code'] == 'unknown-command', reply
        assert pane()['pid'] == pid

        # Real zor observes the pane through the control socket only.
        observer = subprocess.Popen(
            [ZOR, '--rules', str(rules), '--agent', 'test', 'observe', '--socket', str(control), '--pane', '1', '--pid', str(pid)],
            env=env, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        os.set_blocking(observer.stdout.fileno(), False)
        reports = []

        def read_reports():
            try:
                chunk = observer.stdout.read()
            except (TypeError, BlockingIOError):
                chunk = None
            if chunk:
                reports.extend(line for line in chunk.splitlines() if line)
            return reports

        until(lambda: any('state=working' in line for line in read_reports()), 'zor reported working from fux capture')
        assert all(len(line.encode()) <= 1024 for line in reports)
        before = pane()
        rpc(control, {'command': 'send-keys', 'id': 2, 'pane': 1, 'keys': 'x'})
        until(lambda: 'IDLE' in capture(), 'pane keeps working while observed')
        until(lambda: any('state=idle' in line for line in read_reports()), 'zor reported idle from fux capture and title')
        after = pane()
        assert after['pid'] == pid, 'observation changed the pane process'
        assert after['focused'] == before['focused'], 'observation changed focus'
        assert after['title'] == 'OBS_IDLE'
        # Killing the observer never touches the pane.
        observer.kill()
        observer.wait(timeout=5)
        assert pane()['pid'] == pid
        assert 'IDLE' in capture()
        print('PASS zor observes fux through the control protocol; observer loss and bad clients leave panes untouched', flush=True)
    finally:
        if observer is not None and observer.poll() is None:
            observer.kill()
            observer.wait()
        server.terminate()
        try:
            server.wait(timeout=10)
        except subprocess.TimeoutExpired:
            server.kill()
            server.wait()
            raise
        stderr = server.stderr.read().decode()
        assert 'panicked' not in stderr, stderr

File: tools/xtask/src/scenarios/observer.rs
//! Independent real zor observation over fux's public control protocol.
use crate::support::{
    local::{Root, completed, connect, rpc, until},
    logged::Logged,
    process::{self, OwnedProcess, stop_owned},
};
use anyhow::{Context, Result, ensure};
use serde_json::{Value, json};
use std::{
    collections::BTreeSet,
    fs,
    io::{Read, Write},
    path::Path,
    time::{Duration, Instant},
};

fn wait_report(child: &mut Logged, reports: &mut Vec<Vec<u8>>, marker: &str) -> Result<()> {
    until(Duration::from_secs(10), || {
        reports.extend(child.lines()?);
        ensure!(
            reports.iter().all(|line| line.len() <= 1024),
            "oversized observation report"
        );
        Ok(reports
            .iter()
            .any(|line| String::from_utf8_lossy(line).contains(marker))
            .then_some(()))
    })
}
fn snapshot(child: &mut Logged, check: impl Fn(&Value) -> bool) -> Result<Value> {
    until(Duration::from_secs(10), || {
        for line in child.lines()? {
            let value = serde_json::from_slice(&line)?;
            if check(&value) {
                return Ok(Some(value));
            }
        }
        Ok(None)
    })
}
fn default_state(value: &Value) -> Option<&str> {
    value["observations"]
        .as_array()?
        .iter()
        .find(|e| e["handle"]["workspace"] == "default")?["state"]
        .as_str()
}
fn hup(child: &mut Logged) -> Result<()> {
    ensure!(
        child.child.0.try_wait()?.is_none(),
        "watcher exited before reload"
    );
    nix::sys::signal::kill(
        nix::unistd::Pid::from_raw(i32::try_from(child.child.0.id())?),
        nix::sys::signal::Signal::SIGHUP,
    )?;
    Ok(())
}
pub(super) fn run(binary: &Path, zor: &Path) -> Result<()> {
    let command = "stty raw -echo; printf 'READY\\033]9;4;1;50\\007'; dd bs=1 count=1 >/dev/null 2>&1; printf '\\033[2J\\033[HIDLE\\033]9;4;0;0\\007\\033]2;OBS_IDLE\\007'; dd bs=1 count=1 >/dev/null 2>&1; printf '\\033[2J\\033[HUNKNOWN\\033]2;\\007'; sleep 60";
    let root = Root::new("fo-rs-", &["/bin/sh".into(), "-c".into(), command.into()])?;
    let rules = root.path().join("rules");
    fs::create_dir(&rules)?;
    let source = "id='test'\nprompt_marker='>'\nblock_markers=[]\n[[rules]]\nid='working'\nstate='working'\nregion='progress'\ncontains=['1:50']\nvisible_working=true\n[[rules]]\nid='idle'\nstate='idle'\nregion='title'\ncontains=['OBS_IDLE']\nvisible_idle=true\n";
    fs::write(rules.join("test.toml"), source)?;
    let mut server = root.server(binary)?;
    let control = root.control();
    let mut observer: Option<Logged> = None;
    let mut watcher: Option<Logged> = None;
    let scenario = (|| -> Result<()> {
        let pane = || -> Result<Value> {
            Ok(
                completed(&control, json!({"command":"list","id":1}))?["workspaces"][0]["tabs"][0]
                    ["panes"][0]
                    .clone(),
            )
        };
        let capture = || -> Result<String> {
            Ok(completed(&control, json!({"command":"capture","id":1,"pane":1,"attrs":false,"scrollback":0,"max_bytes":4096}))?["text"].as_str().context("capture text")?.into())
        };
        let first = pane()?;
        let pid = &first["pid"];
        let instance = completed(&control, json!({"command":"list","id":2}))?["instance"].clone();
        ensure!(
            !instance.as_str().context("instance")?.is_empty(),
            "empty instance"
        );
        for mut op in [
            json!({"command":"capture","pane":1,"max_bytes":4096}),
            json!({"command":"send-keys","pane":1,"keys":"must-not-arrive"}),
            json!({"command":"subscribe","events":[]}),
        ] {
            op["id"] = 9.into();
            op["instance"] = "old-server".into();
            let stale = rpc(&control, op)?;
            ensure!(
                stale["status"] == "failed" && stale["error"]["code"] == "conflict",
                "stale observer request accepted: {stale}"
            );
        }
        until(Duration::from_secs(10), || {
            Ok(capture()?.contains("READY").then_some(()))
        })?;
        until(Duration::from_secs(10), || {
            let capture = completed(
                &control,
                json!({"command":"capture","id":1,"pane":1,"max_bytes":4096}),
            )?;
            Ok((capture["progress"] == json!([1, 50])).then_some(()))
        })?;
        let request = json!({"command":"capture","id":3,"instance":instance,"pane":1,"attrs":true,"scrollback":0,"max_bytes":131072});
        let original = completed(&control, request.clone())?;
        ensure!(
            original["rows"] == first["geometry"]["height"]
                && original["columns"] == first["geometry"]["width"],
            "capture geometry differs"
        );
        ensure!(
            original["progress"] == json!([1, 50]) && original["truncated"] == false,
            "capture metadata differs"
        );
        let mut conditional = request.clone();
        conditional["if_revision"] = original["revision"].clone();
        let cached = completed(&control, conditional.clone())?;
        ensure!(
            cached["unchanged"] == true && cached["text"] == "",
            "conditional capture missed cache"
        );
        for preface_bytes in [b"FUZ\n".as_slice(), b"FUXCTL1\n".as_slice()] {
            let mut bad = connect(&control, Instant::now() + Duration::from_secs(3))?;
            bad.set_read_timeout(Some(Duration::from_secs(3)))?;
            bad.set_write_timeout(Some(Duration::from_secs(3)))?;
            bad.write_all(preface_bytes)?;
            let mut preface = [0; 4];
            let count = match bad.read(&mut preface) {
                Ok(n) => n,
                Err(e)
                    if matches!(
                        e.kind(),
                        std::io::ErrorKind::ConnectionReset | std::io::ErrorKind::BrokenPipe
                    ) =>
                {
                    0
                }
                Err(e) => return Err(e.into()),
            };
            ensure!(
                count == 0 || &preface[..count] == b"FUX\n",
                "bad client unexpected preface"
            );
            drop(bad);
        }
        let reply = rpc(&control, json!({"command":"popup","id":5,"argv":["x"]}))?;
        ensure!(
            reply["status"] == "failed" && reply["error"]["code"] == "unknown-command",
            "unknown command accepted"
        );
        ensure!(pane()?["pid"] == *pid, "bad client changed pane");
        let base_command = || {
            let mut c = root.command(zor);
            c.arg("--rules").arg(&rules).args(["--agent", "test"]);
            c
        };
        let watch_once = || -> Result<Value> {
            let mut c = base_command();
            c.args(["watch", "--once", "--runtime"])
                .arg(control.parent().context("control parent")?);
            let out = process::output(c, Duration::from_secs(10), 1024 * 1024)?;
            ensure!(
                out.status.success(),
                "watch once failed: {}",
                String::from_utf8_lossy(&out.stderr)
            );
            Ok(serde_json::from_slice(&out.stdout)?)
        };
        let discovered = watch_once()?;
        ensure!(
            discovered["problems"]
                .as_object()
                .is_some_and(|p| p.is_empty())
                || discovered["problems"]
                    .as_array()
                    .is_some_and(|p| p.is_empty()),
            "discovery problems: {discovered}"
        );
        ensure!(
            discovered["observations"][0]["state"] == "working"
                && discovered["observations"][0]["handle"]["pid"] == *pid
                && discovered["observations"][0]["handle"]["instance"] == instance,
            "discovery mismatch: {discovered}"
        );
        let mut c = root.command(binary);
        c.args(["workspace", "new", "other"]);
        ensure!(
            process::output(c, Duration::from_secs(10), 1024 * 1024)?
                .status
                .success(),
            "second workspace failed"
        );
        let multiple = watch_once()?;
        let observations = multiple["observations"]
            .as_array()
            .context("observations")?;
        let names: BTreeSet<_> = observations
            .iter()
            .filter_map(|e| e["handle"]["workspace"].as_str())
            .collect();
        ensure!(
            names == BTreeSet::from(["default", "other"]),
            "workspace discovery: {multiple}"
        );
        ensure!(
            observations
                .iter()
                .all(|e| e["handle"]["instance"] == instance),
            "mixed instances"
        );
        let duration = multiple["scan_duration_ms"]
            .as_u64()
            .context("scan duration")?;
        ensure!(
            observations.iter().all(|e| e["age_upper_bound_ms"]
                .as_u64()
                .is_some_and(|age| age <= duration)),
            "observation age bound"
        );
        let observe_command = || {
            let mut c = base_command();
            c.args(["observe", "--socket"]).arg(&control).args([
                "--pane",
                "1",
                "--pid",
                &pid.to_string(),
            ]);
            c
        };
        observer = Some(Logged::spawn(observe_command())?);
        let mut reports = Vec::new();
        wait_report(
            observer.as_mut().context("observer")?,
            &mut reports,
            "state=working",
        )?;
        let before = pane()?;
        completed(
            &control,
            json!({"command":"send-keys","id":2,"pane":1,"keys":"x"}),
        )?;
        until(Duration::from_secs(10), || {
            Ok(capture()?.contains("IDLE").then_some(()))
        })?;
        let changed = completed(&control, conditional)?;
        ensure!(
            changed["unchanged"] == false
                && changed["title"] == "OBS_IDLE"
                && changed["progress"].is_null(),
            "changed metadata: {changed}"
        );
        wait_report(
            observer.as_mut().context("observer")?,
            &mut reports,
            "state=idle",
        )?;
        let after = pane()?;
        ensure!(
            after["pid"] == *pid
                && after["focused"] == before["focused"]
                && after["title"] == "OBS_IDLE",
            "observer changed pane"
        );
        ensure!(
            watch_once()?["observations"][0]["state"] == "idle",
            "watch missed idle"
        );
        observer.as_mut().context("observer")?.kill()?;
        ensure!(
            pane()?["pid"] == *pid && capture()?.contains("IDLE"),
            "observer loss changed pane"
        );
        let mut c = base_command();
        c.args(["watch", "--runtime"])
            .arg(control.parent().context("control parent")?);
        watcher = Some(Logged::spawn(c)?);
        let watching = watcher.as_mut().context("watcher")?;
        let baseline = snapshot(watching, |v| v["rules_generation"] == 1)?;
        ensure!(
            default_state(&baseline) == Some("idle"),
            "initial watcher state"
        );
        fs::write(rules.join("test.toml"), "malformed[")?;
        hup(watching)?;
        let rejected = snapshot(watching, |v| {
            v["problems"].get("rules").is_some()
                || v["problems"]
                    .as_array()
                    .is_some_and(|p| p.contains(&json!("rules")))
        })?;
        ensure!(
            rejected["rules_generation"] == 1 && default_state(&rejected) == Some("idle"),
            "bad reload discarded old rules"
        );
        fs::write(
            rules.join("test.toml"),
            source
                .replace("id='idle'\nstate='idle'", "id='idle'\nstate='blocked'")
                .replace("visible_idle=true", "visible_blocker=true"),
        )?;
        hup(watching)?;
        let reloaded = snapshot(watching, |v| v["rules_generation"] == 2)?;
        ensure!(
            default_state(&reloaded) == Some("blocked")
                && reloaded["problems"].get("rules").is_none()
                && !reloaded["problems"]
                    .as_array()
                    .is_some_and(|p| p.contains(&json!("rules"))),
            "reload failed: {reloaded}"
        );
        ensure!(pane()?["pid"] == *pid, "reload restarted pane");
        observer = Some(Logged::spawn(observe_command())?);
        reports.clear();
        wait_report(
            observer.as_mut().context("observer")?,
            &mut reports,
            "state=blocked",
        )?;
        completed(
            &control,
            json!({"command":"send-keys","id":8,"pane":1,"keys":"x"}),
        )?;
        until(Duration::from_secs(10), || {
            Ok(capture()?.contains("UNKNOWN").then_some(()))
        })?;
        wait_report(
            observer.as_mut().context("observer")?,
            &mut reports,
            "state=none",
        )?;
        snapshot(watching, |v| default_state(v) == Some("unknown"))?;
        ensure!(pane()?["pid"] == *pid, "unknown restarted pane");
        Ok(())
    })();
    let mut errors = stop_owned(&mut [
        watcher
            .as_mut()
            .map(|c| &mut c.child.0 as &mut dyn OwnedProcess),
        observer
            .as_mut()
            .map(|c| &mut c.child.0 as &mut dyn OwnedProcess),
        Some(&mut server.child),
    ]);
    if let Err(error) = scenario {
        errors.push(format!("{error:#}"));
    }
    match server.diagnostic() {
        Ok(stderr) if stderr.contains("panicked") => errors.push(stderr),
        Err(error) => errors.push(format!("{error:#}")),
        _ => {}
    }
    ensure!(errors.is_empty(), "{}", errors.join("; "));
    println!(
        "PASS zor observes fux through the control protocol; observer loss and bad clients leave panes untouched"
    );
    Ok(())
}
